### PR TITLE
Text2d visibility fix

### LINKED ISSUE #9676: Text2d is broken
## Bevy version

main, since #9236

## What you did

`cargo run --example text2d`

## What went wrong

No text shows up.

## Additional information

This is a known issue, but not yet documented here on GitHub. I'm posting here so there's a place for discussion and to point to in other issues/PRs and so that it's not forgotten.

From discord:

> the queue_sprites is checking against visible_entities in order to determine whether it renders each sprite and since we are using a sparseset for sprites the text2d individual glyphs get assigned new entity ids that don't match the original Text2dBundle so they are considered not visible. Ideally we have a better way of storing text glyphs for both ui and 2d (and eventually 3d) so we don't have to spawn a bunch of empty entities. 

> (this doesn't currently affect UI text as it is clipped later rather than using visible_entities)

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Text drawn with `Text2d` isn't visible in main. 

fixes #9676

## Solution

Add an `ExtractedSpriteBatches` resource that allows the extraction of sprites in batches that all share the same entity for visibility checks. 

In `extract_text2d_sprite` extract to `ExtractedSpriteBatches` rather than `ExtractedSprites`. 

This is a bit of a hack but it's harmless and minimal. 

My first attempt had three buffers:
1. Set of all `Text2d` entities with their  `Entity`, `Transform`, and a range into buffer 2.
2. `Vec` of `ExtractedGlyphSections`. Each `ExtractedGlyphSection` has a handle for the font's atlas texture, color, and a range of indices into buffer 3.
3. `Vec` of `ExtractedGlyphs`. Each `ExtractedGlyph` has a position (relative to the transform) and a `Rect` containing coords of the glyph in the font's atlas texture.

I struggled to get it to work though. It was almost there but there were some weird texturing issues I couldn't work out even after spending half a day on it: 

Capture-text2d
---

I still don't understand Bevy's rendering internals very well, so it could be anything.

Anyway, I decided to switch approach. This PR is deliberately as simple as I could make it. It fixes `Text2d` rendering and shouldn't introduce any new problems or have any significant performance costs. The implementations of `ExtractedSpriteBatches` and `extract_text2d_sprites` aren't super efficient, but naive benchmarks suggest this `Text2d` is faster than 0.11.2's `Text2d` by about 5-10% (probably entirely due to the improvements in bevy_sprites).

```
cargo run --profile stress-test --features trace_tracy --example many_sprites
```
Capture-many-sprites

(yellow is this PR, red is main)

The main drawback is that `ExtractedSpriteBatches` is exposed as public, I'd rather have kept it private as it's not really suitable for much else except text rendering.

---

## Changelog

* Added `ExtractedSpriteBatches`. 
* Modified `queue_sprites` and `prepare_sprites` so they also add the sprites from `ExtractedSpriteBatches`.
* `extract_text2d_sprite` extracts to `ExtractedSpriteBatches` instead of `ExtractedSprites`.

### CODE REVIEW

File: crates/bevy_sprite/src/render/mod.rs [hunk 1]
```diff
@@ -333,6 +333,12 @@ pub struct ExtractedSprites {
     pub sprites: SparseSet<Entity, ExtractedSprite>,
 }
 
+#[derive(Resource, Default)]
+pub struct ExtractedSpriteBatches {
```
Comment (alice-i-cecile) [inline crates/bevy_sprite/src/render/mod.rs:337]: Docs please?

File: crates/bevy_sprite/src/render/mod.rs [hunk 2]
```diff
@@ -590,6 +597,38 @@ pub fn queue_sprites(
                 });
             }
         }
+
+        for (batch_entity, sprite_entities) in extracted_batches.batches.iter() {
+            if !view_entities.contains(batch_entity.index() as usize) {
```
Comment (alice-i-cecile) [inline crates/bevy_sprite/src/render/mod.rs:602]: This could use a bit of a comment: it's really quite unusual to care about the index of an Entity.

Comment (ickshonpe) [reply crates/bevy_sprite/src/render/mod.rs:602]: The original code wasn't commented either, I'm confident these changes work and don't mess anything up but I don't know enough about the rendering internals to be explaining anything to anyone I think.

**Review by mockersf — APPROVED** [review]

File: crates/bevy_sprite/src/render/mod.rs [hunk 3]
```diff
@@ -333,6 +333,27 @@ pub struct ExtractedSprites {
     pub sprites: SparseSet<Entity, ExtractedSprite>,
 }
 
+/// Allows extraction of multiply sprites for a single entity that all share that entity's id for visibility resolution.
```
Comment (mnmaita) [inline crates/bevy_sprite/src/render/mod.rs:336]:
```suggestion
/// Allows the extraction of multiple sprites sharing the same entity id for visibility resolution.
```

I think there's a typo here (multiply -> multiple). I'm also leaving a suggestion to rephrase the comment, hopefully it still conveys the same meaning.

**Comment by rparrett** [comment]
I'm assuming that this needs a major rework after #9685 and #9903?

**Comment by superdump** [comment]
I prefer #10100  if it fixes the same problem.

**Comment by superdump** [comment]
Closing due to merging #10100 If that PR doesn’t address the problem, you’re welcome to come shout at me. 💜

**Pull request closed without merging**